src/creation/creator.rs

```diff
@@ -0,0 +1,301 @@
+        .unit()
+    }
+
+    fn add_units(creator: &mut Creator, units: &[Unit]) {
+        for unit in units {
+            creator.add_unit(unit);
+        }
+    }
+
+    #[test]
+    fn creates_initial_unit() {
+        let n_members = NodeCount(7);
+        let round = 0;
+        let creator = Creator::new(NodeIndex(0), n_members);
+        assert!(creator.can_create(round));
+        assert!(!creator.can_create(round + 1));
+        let (preunit, parent_hashes) = creator.create_unit(round);
+        assert_eq!(preunit.round(), round);
+        assert_eq!(parent_hashes.len(), 0);
+    }
+
+    #[test]
+    fn creates_unit_with_all_parents() {
+        let n_members = NodeCount(7);
+        let mut creators = creator_set(n_members);
+        let new_units = create_units(creators.iter(), 0);
+        let new_units: Vec<_> = new_units
+            .into_iter()
+            .map(|(pu, _)| preunit_to_unit(pu))
+            .collect();
+        let expected_hashes: Vec<_> = new_units.iter().map(|u| u.hash()).collect();
+        let creator = &mut creators[0];
+        add_units(creator, &new_units);
+        let round = 1;
+        assert!(creator.can_create(round));
+        assert!(!creator.can_create(round + 1));
+        let (preunit, parent_hashes) = creator.create_unit(round);
+        assert_eq!(preunit.round(), round);
+        assert_eq!(parent_hashes, expected_hashes);
+    }
+
+    #[test]
+    fn creates_unit_with_minimal_parents() {
+        let n_members = NodeCount(7);
+        let mut creators = creator_set(n_members);
+        let new_units = create_units(creators.iter().take(5), 0);
+        let new_units: Vec<_> = new_units
+            .into_iter()
+            .map(|(pu, _)| preunit_to_unit(pu))
+            .collect();
+        let expected_hashes: Vec<_> = new_units.iter().map(|u| u.hash()).collect();
+        let creator = &mut creators[0];
+        add_units(creator, &new_units);
+        let round = 1;
+        assert!(creator.can_create(round));
+        assert!(!creator.can_create(round + 1));
+        let (preunit, parent_hashes) = creator.create_unit(round);
+        assert_eq!(preunit.round(), round);
+        assert_eq!(parent_hashes, expected_hashes);
+    }
+
+    #[test]
+    fn cannot_create_unit_below_parents_threshold() {
+        let n_members = NodeCount(7);
+        let mut creators = creator_set(n_members);
+        let new_units = create_units(creators.iter().take(4), 0);
+        let new_units: Vec<_> = new_units
+            .into_iter()
+            .map(|(pu, _)| preunit_to_unit(pu))
+            .collect();
+        let creator = &mut creators[0];
+        add_units(creator, &new_units);
+        let round = 1;
+        assert!(!creator.can_create(round));
+    }
+
+    #[test]
+    fn creates_two_units_when_possible() {
+        let n_members = NodeCount(7);
+        let mut creators = creator_set(n_members);
+        let mut expected_hashes_per_round = Vec::new();
+        for round in 0..=1 {
```

Comment:
Since we number rounds starting at 0 this was more intuitive for me, since it shows the highest round involved. I was on the fence though, so changing.